smarc_gazebo_plugins/smarc_gazebo_plugins/src/UnderwaterSonarSensor.cc

```diff
@@ -0,0 +1,579 @@
+  }
+  scan->set_vertical_count(this->VerticalRangeCount());
+
+  scan->set_range_min(this->RangeMin());
+  scan->set_range_max(this->RangeMax());
+
+  scan->clear_ranges();
+  scan->clear_intensities();
+  entities->clear_data();
+
+  unsigned int rayCount = this->RayCount();
+  unsigned int rangeCount = this->RangeCount();
+  unsigned int verticalRayCount = this->VerticalRayCount();
+  unsigned int verticalRangeCount = this->VerticalRangeCount();
+  double rayAngle = this->AngleResolution();
+
+  /*
+  printf("Vertical rays: %u", verticalRayCount);
+  printf("Vertical ranges: %u", verticalRangeCount);
+  printf("Vertical min angle: %f", this->VerticalAngleMin().Radian());
+  printf("Vertical max angle: %f", this->VerticalAngleMax().Radian());
+  printf("Vertical angle step: %f", this->VerticalAngleResolution());
+  printf("Horizontal rays: %u", rayCount);
+  printf("Horizontal ranges: %u", rangeCount);
+  printf("Horizontal angle step: %f", this->AngleResolution());
+  */
+  GZ_ASSERT(verticalRayCount == 2 && verticalRangeCount == 1,
+          "Vertical ray count needs to be 2 for angle interpolation");
+  GZ_ASSERT(rayCount == rangeCount,
+          "Horizontal ray count needs to be the same as ranges");
+
+  // interpolate in horizontal direction
+  for (unsigned int i = 0; i < rangeCount-1; ++i) // TODO: fix the last one eventually
+  {
+    double angle, range, range_right, range_above, intensity;
+    range = this->dataPtr->laserShape->GetRange(this->RayCount() + i);
+    range_right = this->dataPtr->laserShape->GetRange(this->RayCount() + i + 1);
+    range_above = this->dataPtr->laserShape->GetRange(i);
+
+	ignition::math::Vector3d p_right(0.0, range_right*sin(rayAngle), range_right*cos(rayAngle)-range);
+	ignition::math::Vector3d p_above(range_above*sin(rayAngle), 0.0, range_above*cos(rayAngle)-range); // not certain that rayAngle is correct here
+	p_right.Normalize();
+	p_above.Normalize();
+	ignition::math::Vector3d normal = p_right.Cross(p_above);
+	normal.Normalize();
+	angle = acos(fabs(normal.Z()));
+
+	intensity = this->dataPtr->laserShape->GetRetro(this->RayCount() + i);
+    physics::RayShapePtr ray = physics::SemanticMultiRayShape::StaticGetRay(this->dataPtr->laserShape, this->RayCount() + i);
+    double _dist;
+    std::string _entity;
+    ray->GetIntersection(_dist, _entity);
+    //printf("%s", _entity.c_str());
+
+    // Mask ranges outside of min/max to +/- inf, as per REP 117
+    if (range >= this->RangeMax())
+    {
+      range = IGN_DBL_INF;
+    }
+    else if (range <= this->RangeMin())
+    {
+      range = -IGN_DBL_INF;
+    }
+    else if (this->noises.find(RAY_NOISE) !=
+             this->noises.end())
+    {
+      // currently supports only one noise model per laser sensor
+	  range = this->noises[RAY_NOISE]->Apply(range);
+      range = ignition::math::clamp(range,
+              this->RangeMin(), this->RangeMax());
+    }
+
+	// TODO: make these into proper parameters
```

Comment:
@lofux This is where we can add a more reasonable sonar equation (for planes such as the sea floor).

Reply:
(see lines below this comment)